packages/offchain-messages/src/signatures.ts

```diff
@@ -9,6 +15,118 @@ import { OffchainMessageEnvelope } from './envelope';
  */
 export type FullySignedOffchainMessageEnvelope = NominalType<'offchainMessageEnvelopeSignedness', 'fullySigned'>;
 
+/**
+ * Given an array of `CryptoKey` objects which are private keys pertaining to addresses that are
+ * required to sign an offchain message, this method will return a new signed offchain message
+ * envelope of type {@link OffchainMessageEnvelope}.
+ *
+ * Though the resulting message might be signed by all required signers, this function will not
+ * assert that it is. A partially signed message is not complete, but can be serialized and
+ * deserialized.
+ *
+ * @example
+ * ```ts
+ * import { generateKeyPair } from '@solana/keys';
+ * import { partiallySignOffchainMessageEnvelope } from '@solana/offchain-messages';
+ *
+ * const partiallySignedOffchainMessage = await partiallySignOffchainMessageEnvelope(
+ *     [myPrivateKey],
+ *     offchainMessageEnvelope,
+ * );
+ * ```
+ *
+ * @see {@link signOffchainMessageEnvelope} if you want to assert that the message is signed by all
+ * its required signers after signing.
+ */
+export async function partiallySignOffchainMessageEnvelope<TOffchainMessageEnvelope extends OffchainMessageEnvelope>(
+    keyPairs: CryptoKeyPair[],
+    offchainMessageEnvelope: TOffchainMessageEnvelope,
+): Promise<TOffchainMessageEnvelope> {
+    let newSignatures: Record<Address, SignatureBytes> | undefined;
+    let unexpectedSigners: Set<Address> | undefined;
+
+    await Promise.all(
+        keyPairs.map(async keyPair => {
+            const address = await getAddressFromPublicKey(keyPair.publicKey);
+            const existingSignature = offchainMessageEnvelope.signatures[address];
+
+            // Check if the address is expected to sign the message
+            if (existingSignature === undefined) {
+                // address is not an expected signer for this message
+                unexpectedSigners ||= new Set();
+                unexpectedSigners.add(address);
+                return;
+            }
```

Comment:
This is a copy-paste of the transaction signing code. This code doesn't _actually_ check the message content to read the required signers; it trusts that the message envelope `signatures` map was compiled from the message preamble faithfully. Thoughts on this?